scenario/state.py

```diff
@@ -609,6 +610,73 @@ class Mount(_DCBase):
     src: Union[str, Path]
 
 
+def _now_utc():
+    return datetime.datetime.now(tz=datetime.timezone.utc)
+
+
+# Ideally, this would be a subclass of pebble.Notice, but we want to make it
+# easier to use in tests by providing sensible default values, but there's no
+# default key value, so that would need to be first, and that's not the case
+# in pebble.Notice, so it's easier to just be explicit and repetitive here.
+@dataclasses.dataclass(frozen=True)
+class PebbleNotice(_DCBase):
+    key: str
+    """The notice key, a string that differentiates notices of this type.
+
+    This is in the format ``domain/path``; for example:
+    ``canonical.com/postgresql/backup`` or ``example.com/mycharm/notice``.
+    """
+
+    notice_id: str = dataclasses.field(default_factory=lambda: str(uuid4()))
+    """Unique ID for this notice."""
+
+    user_id: Optional[int] = None
+    """UID of the user who may view this notice (None means notice is public)."""
+
+    type: Union[pebble.NoticeType, str] = pebble.NoticeType.CUSTOM
+    """Type of the notice."""
+
+    first_occurred: datetime.datetime = dataclasses.field(default_factory=_now_utc)
+    """The first time one of these notices (type and key combination) occurs."""
+
+    last_occurred: datetime.datetime = dataclasses.field(default_factory=_now_utc)
+    """The last time one of these notices occurred."""
+
+    last_repeated: datetime.datetime = dataclasses.field(default_factory=_now_utc)
+    """The time this notice was last repeated.
+
+    See Pebble's `Notices documentation <https://github.com/canonical/pebble/#notices>`_
+    for an explanation of what "repeated" means.
+    """
+
+    occurrences: int = 1
+    """The number of times one of these notices has occurred."""
+
+    last_data: Dict[str, str] = dataclasses.field(default_factory=dict)
```

Comment:
Would it gain us anything (or even be sensible) to inherit from `pebble.Notice`, even if we have to override the defaults on (more than) half the fields?

Reply:
I started out trying to do that (it feels like it would be worth doing with `Port` and the the status class, too). The problem is that I do think it's worth having default values in Scenario to avoid having to write out a lot of boilerplate in the tests, but if we have the two classes like:

```
@dataclasses.dataclass
class A:  # ops.pebble.Notice
    id: str
    user_id: int|None
    type: str
    key: str


@dataclasses.dataclass
class B:  # scenario.PebbleNotice
    id: str = 'example'
    user_id: int|None = None
    type: str = "custom"
    key: str
```

Then this errors out because there's a non-default argument (`key`) after ones with defaults. We can fix that by moving the key up to be the first attribute, but then the two classes have different `__init__` argument ordering (but the same names) and that feels like it's a trap waiting for someone to fall into.

In `ops`, `Notice`s are created with keyword args, but the dataclass isn't marked as `kw_only` (and can't be because (a) that would break backwards compatibility, and (b) that needs Python 3.10).

I think it is sensible, and it would avoid having the `_to_pebble_notice` method that generates an `ops.pebble.Notice` from a `scenario.PebbleNotice` so that it can be used in the mock pebble client (`_to_pebble_notice is probably a bad name because it's `PebbleNotice._to_pebble_notice`, maybe should be `_to_ops_pebble_notice`?).

Overall, it felt like it was cleaner to just separate them out.

Reply:
Thanks -- yeah, I think you're right. It's cleaner to keep all of these separate so we can independently move the types forward, and have the `state.Foo` dataclasses be customized/tailored for Scenario usage, where that doesn't make sense for the ops ones.

I'd go for `scenario.PebbleNotice._to_ops_notice` I think.